README.md
# Sonoff LAN control from Home Assistant

[![hacs_badge](https://img.shields.io/badge/HACS-Custom-orange.svg)](https://github.com/custom-components/hacs)
[![Donate](https://img.shields.io/badge/donate-Coffee-yellow.svg)](https://www.buymeacoffee.com/AlexxIT)
[![Donate](https://img.shields.io/badge/donate-Yandex-red.svg)](https://money.yandex.ru/to/41001428278477)

- [Readme in Russian](https://github.com/AlexxIT/SonoffLAN/blob/master/README_ru.md)

Home Assistant Custom Component for control **eWeLink** (Sonoff) devices over Local Network (LAN).

**Support only devices with firmware v3+**. LAN should support **Multicast** traffic.

Supporting firmware v2 **in development** ([read more](https://github.com/AlexxIT/SonoffLAN/issues/31)). Unfortunately I do not have such devices.

Pros:

- work with original eWeLink/Sonoff firmware, no need to flash devices
- work over local network (LAN), no Cloud Server dependency
- work with devices without DIY-mode
- work with devices in DIY-mode
- support single and multi-channel devices
- support TH and POW device attributes
- support Sonoff RF Bridge 433 for receive and send commands
- instant device state update with Multicast
- (optional) load devices list from eWeLink Servers (with names, apikey/devicekey and device_class) and save it locally
- (optional) change device type (switch, light or fan)
- (optional) set multi-channel device as one light with brightness control

**Component review from DrZzs**

[![Component review from DrZzs](https://img.youtube.com/vi/DsTqOlrQQ1k/0.jpg)](https://www.youtube.com/watch?v=DsTqOlrQQ1k)

There is another great component by [@peterbuga](https://github.com/peterbuga/HASS-sonoff-ewelink), that works with cloud servers.

Thanks to these people [@beveradb](https://github.com/beveradb/sonoff-lan-mode-homeassistant), [@mattsaxon](https://github.com/mattsaxon/sonoff-lan-mode-homeassistant) for researching the local Sonoff protocol.

## Tested Devices

- [Sonoff Basic](https://www.itead.cc/sonoff-wifi-wireless-switch.html) fw 3.0.1
- [Sonoff Mini](https://www.itead.cc/sonoff-mini.html) (no need use DIY-mode) fw 3.3.0
- [Sonoff TH](https://www.itead.cc/sonoff-th.html) (show temperature and humidity)
- [Sonoff 4CH Pro R2](https://www.itead.cc/sonoff-4ch-pro.html) fw 3.3.0
- [Sonoff Pow R2](https://www.itead.cc/sonoff-pow-r2.html) (show power consumption)
- [Sonoff Micro](https://www.itead.cc/sonoff-micro-5v-usb-smart-adaptor.html) fw 3.4.0
- [Sonoff RF Bridge 433](https://www.itead.cc/sonoff-rf-bridge-433.html) (receive and send commands) fw 3.3.0
- [Sonoff D1](https://www.itead.cc/sonoff-d1-smart-dimmer-switch.html) (dimmer with brightness control) fw 3.4.0
- [Sonoff iFan02](https://www.itead.cc/ru/sonoff-ifan02-wifi-smart-ceiling-fan-with-light.html) (light and fan with speed control) fw 3.3.0
- [Sonoff iFan03](https://www.itead.cc/sonoff-ifan03-wifi-ceiling-fan-light-controller.html) (light and fan with speed control) 3.4.0
- [Sonoff S20](https://www.itead.cc/smart-socket.html)
- [Sonoff S26](https://www.itead.cc/sonoff-s26-wifi-smart-plug.html)
- [Sonoff S31](https://www.itead.cc/sonoff-s31.html) (show power consumption)
- [Sonoff S55](https://www.itead.cc/sonoff-s55.html)
- [Sonoff SV](https://www.itead.cc/ru/sonoff-sv.html) fw 3.0.1
- [Sonoff T4EU1C](https://www.itead.cc/sonoff-t4eu1c-wi-fi-smart-single-wire-wall-switch.html)
- [Sonoff 5V DIY](https://www.aliexpress.com/item/32818293817.html)
- [MiniTiger Wall Switch](https://www.aliexpress.com/item/33016227381.html) (I have 8 without zero-line) fw 3.3.0

## Config Examples

Minimum config:

```yaml
sonoff:
  username: mymail@gmail.com
  password: mypassword
```

or

```yaml
sonoff:
  username: +910123456789  # important to use country code
  password: mypassword
```

Advanced config:

```yaml
sonoff:
  username: mymail@gmail.com
  password: mypassword
  reload: always  # update device list every time HA starts
  default_class: light  # changes the default class of all devices from switch to light
  devices:
    1000abcdefg:
      device_class: light  # changes the default class of the device from switch to light
```

Devices can be set manually, without connecting to Cloud Servers. But in this case, you need to know the `devicekey` for each device.

```yaml
sonoff:
  devices:
    1000abcdefg:
      devicekey: f9765c85-463a-4623-9cbe-8d59266cb2e4
```

Examples of using `device_class`:

```yaml
sonoff:
  username: mymail@gmail.com
  password: mypassword
  reload: once
  devices:
    1000abcde0: # corridor light
      device_class: light
    1000abcde1: # children's light (double switch, one light entity)
      device_class:
      - light: [1, 2]
    1000abcde2: # toilet light and fan (double switch)
      device_class: [light, fan]
    1000abcde3: # bedroom light and backlight (double switch)
      device_class: [light, light]
    1000abcde4: # hall three light zones Sonoff 4CH
      device_class:
      - light # zone 1 (channel 1)
      - light # zone 2 (channel 2)
      - light: [3, 4] # zone 3 (channels 3 and 4)
```

Minimum config for devices only in DIY mode:

```yaml
sonoff:
```

## Sonoff RF Bridge 433

Component will create only one entity per RF Bridge - `remote.sonoff_1000abcdefg`. Entity RF Buttons or RF Sensors are not created!

You can receive signals from RF Buttons and RF Sensors through an event `sonoff.remote`. And send signals using the service `remote.send_command`.

Although the component supports training, it is recommended to train RF Buttons through the eWeLink application.

When a command is received, the event `sonoff.remote` is generated with a button number and response time (in UTC, sends the device).

`command` - number of the button in the eWeLink application.

Example for receive RF signal via [Automation](https://www.home-assistant.io/integrations/automation/):

```yaml
automation:
- alias: Receive RF Button1
  trigger:
    platform: event
    event_type: sonoff.remote
    event_data:
      name: Button1  # button/sensor name in eWeLink application
  action:
    service: homeassistant.toggle
    entity_id: switch.sonoff_1000abcdefg
```

Example for send RF signal via [Script](https://www.home-assistant.io/integrations/script/):

```yaml
script:
  send_button1:
    alias: Send RF Button1
    sequence:
    - service: remote.send_command
      data:
        entity_id: remote.sonoff_1000abcdefg
        command: Button1  # button name in eWeLink application
```

## Sonoff TH и Pow

Temperature, humidity and other parameters of the devices are stored in their attributes. They can be displayed through [Template](https://www.home-assistant.io/integrations/template/)-sensor.

```yaml
sensor:
- platform: template
  sensors:
    temperature_purifier:
      friendly_name: Temperature
      device_class: temperature
      value_template: "{{ state_attr('switch.sonoff_1000abcdefg', 'temperature') }}"
    humidity_purifier:
      friendly_name: Humidity
      device_class: humidity
      value_template: "{{ state_attr('switch.sonoff_1000abcdefg', 'humidity') }}"
```

## Parameters:

- **reload** - *optional*  
  `always` - load device list every time HA starts  
  `once` - (default) download device list once
- **default_class** - *optional*, default `switch`, overrides default device type of all devices
- **device_class** - *optional*, overrides device type (default all **sonoff** devices are displayed as `default_class`). May be a string or an array of strings (for multi-channel switches). Supports types: `light`, `fan`, `switch`, `remote` (only for *Sonoff RF Bridge 433*).


## Work with Cloud Servers

With `username` and` password` in the config (optional) - component loads list of devices from eWeLink Servers and save it in the file `/config/.sonoff.json` (hidden file).

The component does not make other requests to servers.

The list will be loaded only once. At the next start, the list will be loaded from the local file. When you have new **eWeLink** devices - manually delete the file and reboot the HA.

With `reload: always` in the config - the list will be loaded from servers at each start.

The list will be loaded from the local file even if you remove `username` and `password` from the settings.

## Getting devicekey manually

1. Put the device in setup mode
2. Connect to the Wi-Fi network `ITEAD-10000`, password` 12345678`
3. Open in browser `http://10.10.7.1/device`
4. Copy `deviceid` and `apikey` (this is `devicekey`)
5. Connect to your Wi-Fi network and setup Sonoff via the eWeLink app

## Demo

**Sonoff 4CH Pro R2**, configured as a single light source with brightness control.

[![Control Sonoff Devices with eWeLink firmware over LAN from Home Assistant](https://img.youtube.com/vi/X7PcYfDy57A/0.jpg)](https://www.youtube.com/watch?v=X7PcYfDy57A)

## HACS Support

![Support HACS](hacs.png)

## Common problems

**Devices are not displayed**

1. Currently only supported devices with firmware v3+
2. Common problems with Multicast:
   - two routers
   - docker with port forwarding
   - virtual machine with port forwarding
   - virtualbox
   - linux firewall
   - linux network driver

## Component Debugging

Add to your `configuration.yaml`:

```yaml
logger:
  default: info
  logs:
    custom_components.sonoff: debug
```

Only devices with firmware 3 and higher are supported.

All unknown devices with command `switch` support will be added as `switch`.

All other unknown devices will be added as `binary_sensor` (always `off`). The full state of the device is displayed in its attributes.

The component adds the service `sonoff.send_command` to send low-level commands.

Example service params to single switch:

```yaml
device: 1000123456
command: switch
switch: 'on'
```

Example service params to multi-channel switch:

```yaml
device: 1000123456
command: switches
switches: [{outlet: 0, switch: 'off'}]
```

Example service params to dimmer:

```yaml
device: 1000123456
command: dimmable
switch: 'on'
brightness: 50
mode: 0
```

## Useful Links

- https://github.com/peterbuga/HASS-sonoff-ewelink
- https://github.com/beveradb/sonoff-lan-mode-homeassistant
- https://github.com/mattsaxon/sonoff-lan-mode-homeassistant
- https://blog.ipsumdomus.com/sonoff-switch-complete-hack-without-firmware-upgrade-1b2d6632c01
- https://github.com/itead/Sonoff_Devices_DIY_Tools/blob/master/SONOFF%20DIY%20MODE%20Protocol%20Doc%20v1.4.md

CHANGELOG.md
# Changelog

## 0.2.16 - 2020-02-25

### Added

- `default_class` config option 

## 0.2.15 - 2020-02-24

### Changed

- The component is completely rewritten to asynchronous work.

## 0.2.14 - 2020-02-22

### Added

- Support Sonoff iFan02 and iFan03 (light and fan with speed control)

## 0.2.13 - 2020-02-21

### Added

- Load friendly names for **RF Bridge** buttons from cloud servers
- **RF Bridge** displays last pressed button in attributes
- Send **RF Bridge** command by button name
- Load friendly names for multi-channel switch child buttons from cloud servers

You must manually delete the `.sonoff.json` file to update names.

## 0.2.12 - 2020-02-20

### Added

- Service `sonoff.send_command` to send low-level commands.

### Changed

- Unknown devices will be added as binary_sensor. For debugging purposes

## 0.2.11 - 2020-02-14

### Added

- Support login with phone number

## 0.2.10 - 2020-02-13

### Fixed

- Some users devices send updates several times
- Some users had update errors at startup

## 0.2.9 - 2020-02-12

### Added

- Support Sonoff D1 (dimmer)
- Automatic detection `device_class` for popular devices

## 0.2.8 - 2020-01-25

### Changed

- Add config check (fix only numbers in password or deviceid)

## 0.2.7 - 2020-01-25

### Fixed

- Support non Euro eWeLink server
- Fix readme about `reload: always` default value

## 0.2.6 - 2020-01-22

### Added

- rssi device attribute

## 0.2.5 - 2020-01-21

### Added

- Support Sonoff Pow

## 0.2.4 - 2020-01-18

### Added

- Support Sonoff in DIY mode

## 0.2.3 - 2020-01-18

### Added

- Support Sonoff RF Bridge 433

## 0.2.2 - 2020-01-18

### Fixed

- Support Apple HomeKit

## 0.2.1 - 2020-01-15

### Added

- Support [HACS](https://hacs.xyz/) as Custom Repo

## 0.2.0 - 2020-01-12

### Added

- Support download devices list from [eWeLink](https://www.ewelink.cc/en/) servers
- Support control `device_class` (`switch`, `light`, `fan`)
- Support multichannel devices as single `light` with brightness controll

# Changed

- `apikey` renamed to `devicekey`

## 0.1.1 - 2019-11-20

### Added

- Support **Sonoff TH** (temperature and humidity)

## 0.1.0 - 2019-11-16

### Added

- Support control **Sonoff** (eWeLink) devices over LAN

custom_components/sonoff/utils.py
import base64
import hashlib
import hmac
import json
import logging
import os
import time
import uuid
from base64 import b64decode
from base64 import b64encode
from typing import Optional

from aiohttp import ClientSession

from Crypto.Cipher import AES
from Crypto.Hash import MD5
from Crypto.Random import get_random_bytes
from Crypto.Util.Padding import pad
from Crypto.Util.Padding import unpad

_LOGGER = logging.getLogger(__name__)


def _params(**kwargs):
    """Generate params for sonoff API. Pretend mobile application."""
    return {
        **kwargs,
        'version': 6,
        'ts': int(time.time()),
        'nonce': int(time.time() * 100000),
        'appid': 'oeVkj2lYFGnJu5XUtWisfW4utiN4u9Mq',
        'imei': str(uuid.uuid4()),
        'os': 'android',
        'model': '',
        'romVersion': '',
        'appVersion': '3.12.0'
    }


def load_cache(filename: str) -> Optional[dict]:
    """Load device list from file."""
    if os.path.isfile(filename):
        with open(filename, 'rt', encoding='utf-8') as f:
            return json.load(f)
    return None


def save_cache(filename: str, data: dict):
    """Save device list to file."""
    with open(filename, 'w', encoding='utf-8') as f:
        json.dump(data, f, ensure_ascii=False, separators=(',', ':'))


async def load_devices(username: str, password: str, session: ClientSession):
    """Load device list from ewelink servers."""

    # add a plus to the beginning of the phone number
    if '@' not in username and not username.startswith('+'):
        username = f"+{username}"

    params = _params(email=username, password=password) \
        if '@' in username else \
        _params(phoneNumber=username, password=password)

    hex_dig = hmac.new(b'6Nz4n0xA8s8qdxQf2GqurZj2Fs55FUvM',
                       json.dumps(params).encode(),
                       digestmod=hashlib.sha256).digest()
    headers = {'Authorization': "Sign " + base64.b64encode(hex_dig).decode()}

    r = await session.post('https://eu-api.coolkit.cc:8080/api/user/login',
                           headers=headers, json=params)
    resp = await r.json()

    if 'region' not in resp:
        info = 'email' if '@' in username else 'phone'
        _LOGGER.error(f"Login with {info} error: {resp}")
        return None

    region = resp['region']
    if region != 'eu':
        _LOGGER.debug(f"Redirect to region: {region}")
        r = await session.post(
            f"https://{region}-api.coolkit.cc:8080/api/user/login",
            headers=headers, json=params)
        resp = await r.json()

    headers = {'Authorization': "Bearer " + resp['at']}
    params = _params(apiKey=resp['user']['apikey'], lang='en', getTags=1)
    r = await session.get(
        f"https://{region}-api.coolkit.cc:8080/api/user/device",
        headers=headers, params=params)
    resp = await r.json()

    if resp['error'] == 0:
        return resp['devicelist']

    return None


def encrypt(payload: dict, devicekey: str):
    devicekey = devicekey.encode('utf-8')

    hash_ = MD5.new()
    hash_.update(devicekey)
    key = hash_.digest()

    iv = get_random_bytes(16)
    plaintext = json.dumps(payload['data']).encode('utf-8')

    cipher = AES.new(key, AES.MODE_CBC, iv=iv)
    padded = pad(plaintext, AES.block_size)
    ciphertext = cipher.encrypt(padded)

    payload['encrypt'] = True
    payload['iv'] = b64encode(iv).decode('utf-8')
    payload['data'] = b64encode(ciphertext).decode('utf-8')

    return payload


def decrypt(payload: dict, devicekey: str):
    try:
        devicekey = devicekey.encode('utf-8')

        hash_ = MD5.new()
        hash_.update(devicekey)
        key = hash_.digest()

        encoded = ''.join([payload[f'data{i}'] for i in range(1, 4, 1)
                           if f'data{i}' in payload])

        cipher = AES.new(key, AES.MODE_CBC, iv=b64decode(payload['iv']))
        ciphertext = b64decode(encoded)
        padded = cipher.decrypt(ciphertext)
        return unpad(padded, AES.block_size)

    except:
        return None


UIIDS = {}
TYPES = {}


def init_device_class(default_class: str = 'switch'):
    switch1 = default_class
    switch2 = [default_class, default_class]
    switch3 = [default_class, default_class, default_class]
    switch4 = [default_class, default_class, default_class, default_class]
    switchx = [default_class]

    UIIDS.update({
        1: switch1,
        2: switch2,
        3: switch3,
        4: switch4,
        5: switch1,
        6: switch1,
        7: switch2,
        8: switch3,
        9: switch4,
        28: 'remote',  # Sonoff RF Brigde 433
        29: switch2,
        30: switch3,
        31: switch4,
        34: ['light', {'fan': [2, 3, 4]}],  # Sonoff iFan02 and iFan03
        44: 'light',  # Sonoff D1
        77: switchx,  # Sonoff Micro
        78: switchx,
        81: switch1,
        82: switch2,
        83: switch3,
        84: switch4,
        107: switchx
    })

    TYPES.update({
        'plug': switch1,  # Basic, Mini
        'enhanced_plug': switch1,  # Sonoff Pow R2?
        'th_plug': switch1,  # Sonoff TH?
        'strip': switch4,  # 4CH Pro R2, Micro!, iFan02!
        'light': 'light',  # D1
        'rf': 'remote',  # RF Bridge 433
        'fan_light': ['light', 'fan'],  # iFan03
    })


def guess_device_class(config: dict):
    """Get device_class from uiid (from eWeLink Servers) or from zeroconf type.

    Sonoff iFan02 and iFan03 both have uiid 34. But different types (strip and
    fan_light) and different local API for each type. Without uiid iFan02 will
    be displayed as 4 switches.
    """
    uiid = config.get('uiid')
    type_ = config.get('type')
    return UIIDS.get(uiid) or TYPES.get(type_)
